### PR TITLE
WIP: DatetimeArray+TimedeltaArray

### PR DESCRIPTION
**Pull request by jbrockmendel**
I'll be traveling later this week, so trying to get a few things online before disappearing.

So far this just implements `take` and `concat_same_type`.  Tests are taken from an older mothballed PR, so are pretty duplicative, need to be cleaned up.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Quick comment: IMO it is not needed to write tests for basic functioning of eg `_concat_same_type` or `take`, this is already covered by the base extension tests. 
(for `take`, there is of course some specific handling of the fill_value, which should be tested specifically for datetime-likes)

**Comment by jorisvandenbossche** [comment]
@jbrockmendel What is the status of this, or how does it related to the other PRs? Is this still something you plan to continue working on (so worth to review), or not?

**Comment by jbrockmendel** [comment]
This is essentially orthogonal to the other PRs ATM.  It has been in hold in part because I expect you would be -1 on implementing any parts of the EA interface without implementing all of them.  If that expectation is incorrect, then I would probably add `__repr__` and other easily-reviewed needed-eventually methods.

**Comment by jorisvandenbossche** [comment]
My expectation was that you would further expand this PR to cover the full EA interface, but that is not the plan?

**Comment by jbrockmendel** [comment]
> My expectation was that you would further expand this PR to cover the full EA interface, but that is not the plan?

That was more or less the plan (the "more or less" being that I would have been OK with splitting it into chunks).  Besides the trouble with actually implementing this, we've got pieces of this conversation going on in at least four threads, and this is the one I'm most comfortable with letting lie dormant.

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -132,7 +133,7 @@ def wrapper(self, other):
                 return ops.invalid_comparison(self, other, op)
             else:
                 self._assert_tzawareness_compat(other)
-                result = meth(self, np.asarray(other))
+                result = meth(self, type(self)(other))#np.asarray(other))
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimes.py]: should have been fixed before pushing

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -192,6 +196,107 @@ def astype(self, dtype, copy=True):
             return self._box_values(self.asi8)
         return super(DatetimeLikeArrayMixin, self).astype(dtype, copy)
 
+    # ------------------------------------------------------------------
+    # ExtensionArray Interface
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: is this a todo list? or documentation

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -192,6 +196,107 @@ def astype(self, dtype, copy=True):
...
+    def _concat_same_type(cls, to_concat):
+        # for TimedeltaArray and PeriodArray; DatetimeArray overrides
+        freqs = {x.freq for x in to_concat}
+        assert len(freqs) == 1
+        freq = list(freqs)[0]
+        values = np.concatenate([x.asi8 for x in to_concat])
+        return cls._simple_new(values, freq=freq)
+
+    def copy(self, deep=False):
+        # TODO: should `deep` determine whether we copy self.asi8?
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: yes like
```
values = self.asi8
if deep:
    values = values.copy()
....
```

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -192,6 +196,107 @@ def astype(self, dtype, copy=True):
...
+        if is_datetime64tz_dtype(self):
+            return type(self)(self.asi8.copy(), tz=self.tz, freq=self.freq)
+        return type(self)(self.asi8.copy(), freq=self.freq)
+
+    # Following how PeriodArray does this
+    # TODO: ignoring `type`?
+    def view(self, dtype=None, type=None):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: why is this so complicated? do we really need this method?

File: pandas/core/arrays/datetimes.py [hunk 5]
```diff
@@ -411,6 +412,38 @@ def __iter__(self):
...
+            self._assert_tzawareness_compat(fill_value)
+            fill_value = Timestamp(fill_value).value
+        else:
+            raise ValueError("'fill_value' should be a Timestamp. "
+                             "Got '{got}'.".format(got=fill_value))
+        return fill_value
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        freqs = {x.freq for x in to_concat}
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py]: you do this freq dance a few times, maybe a helper function?

File: pandas/core/series.py [hunk 6]
```diff
@@ -4242,6 +4242,9 @@ def _try_cast(arr, take_fast_path):
                                            not (is_iterator(subarr) or
                                            isinstance(subarr, np.ndarray))):
                 subarr = construct_1d_object_array_from_listlike(subarr)
+            elif type(subarr).__name__ == "DatetimeArray":
```
Comment (jreback) [inline pandas/core/series.py]: woa??

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/series.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/series.py]: Yah this is a kludge that should have gotten changed to an `isinstance(subarr, ABCDatetimeArray)` before pushing

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: TODO list

File: pandas/core/arrays/datetimes.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: I'll take a look.

File: pandas/core/arrays/datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: I implemented this because tests were raising AttributeErrors asking for it.

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: yeah we really *can'* support ``.view`` generally with EA. maybe just leave this on Index.

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: ok sure

**Pull request closed without merging**